Bluetooth: Align error handling of send functions

Vudentz opened the pull request:
This aligns the error handling of send function to never unref the
buffer in place so the caller retain the ownership of the buffer
whenever there is an error.

Vudentz committed:
Bluetooth: Align error handling of send functions

This aligns the error handling of send function to never unref the
buffer in place so the caller retain the ownership of the buffer
whenever there is an error.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: L2CAP: Add note for bt_l2cap_chan_send

This adds a note about buffer ownership and error handling of
bt_l2cap_chan_send so it is aligned with recent changes.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: ISO: Add note for bt_iso_chan_send

This adds a note about buffer ownership and error handling of
bt_iso_chan_send so it is aligned with recent changes.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Thalley commented:
@Vudentz Glad to see the alignment. 

What's the reasoning of having the caller unref the buffer in case of error, rather than unref'ing the buffer in the functions themselves in case of error? The latter of which will be able to make the send functions "fire and forget" rather than "fire and if success, then forget" :D

Vudentz commented:
> @Vudentz Glad to see the alignment.
> 
> What's the reasoning of having the caller unref the buffer in case of error, rather than unref'ing the buffer in the functions themselves in case of error? The latter of which will be able to make the send functions "fire and forget" rather than "fire and if success, then forget" :D

In some cases the caller may want to resent it later in case of an error and while it would be possible to add an extra reference to prevent the buffer to be freed this sort of logic may create some confusion as we experience recently with bugs in att.c, besides there could be error related to the buffer itself like it being too big, etc.

carlescufi commented:
@Vudentz could you please add a note to the release notes?

joerchan reviewed (dismissed):
It think it would be best to include this change in the release notes for clarification.

Vudentz committed:
doc: release: Add note about send changes

This adds a note about the changes made to align the error handling of
the likes of bt_l2cap_chan_send and bt_iso_chan_send.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
> It think it would be best to include this change in the release notes for clarification.

Note as been added.

Vudentz commented:
@jhedberg @joerchan all comments should have been addressed.
